src/supervisor3.erl

```diff
@@ -937,7 +941,8 @@ is_abnormal_termination({shutdown, _}) -> false;
 is_abnormal_termination(_Other)        -> true.
 
 do_restart_delay({RestartType, Delay}, Reason, Child, State) ->
-    case add_restart(State) of
+    IsCleanRetry = Reason =:= {?MODULE, delayed_restart},
+    case maybe_add_restart(State, IsCleanRetry) of
```

Comment:
IsCleanRetry is only used in maybe_add_restart, may be that one can check the Reason directly?